packages/simcore-sdk/src/simcore_sdk/node_ports_common/filemanager.py

```diff
@@ -103,48 +137,167 @@ async def _download_link_to_file(session: ClientSession, url: URL, file_path: Pa
+                _upload_file_part(
+                    session,
+                    file,
+                    index,
+                    index * file_chunk_size,
+                    this_file_chunk_size,
+                    num_urls,
+                    upload_url,
+                    pbar,
+                )
+            )
+        try:
+            results = await logged_gather(*upload_tasks, log=log, max_concurrency=4)
+            part_to_etag = [
+                UploadedPart(number=index + 1, e_tag=e_tag) for index, e_tag in results
+            ]
+            log.info(
+                "Uploaded %s, received %s",
+                f"{file=}",
+                f"{part_to_etag=}",
+            )
+            return part_to_etag
+        except ClientError as exc:
+            raise exceptions.S3TransferError(
+                f"Could not upload file {file}:{exc}"
+            ) from exc
 
 
-async def _upload_file_to_link(
-    session: ClientSession, url: URL, file_path: Path
+async def _complete_upload(
+    session: ClientSession,
+    upload_links: FileUploadSchema,
+    parts: list[UploadedPart],
 ) -> ETag:
-    log.debug("Uploading from %s to %s", file_path, url)
-    file_size = file_path.stat().st_size
+    """completes a potentially multipart upload in AWS
+    NOTE: it can take several minutes to finish, see [AWS documentation](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html)
+    it can take several minutes
+    :raises ValueError: _description_
+    :raises exceptions.S3TransferError: _description_
+    :rtype: ETag
+    """
+    log.debug("completing upload of %s", f"{upload_links=} with {parts=}")
+    async with session.post(
+        upload_links.links.complete_upload,
+        json=jsonable_encoder(FileUploadCompletionBody(parts=parts)),
+    ) as resp:
+        resp.raise_for_status()
+        # now poll for state
+        file_upload_complete_response = parse_obj_as(
+            Envelope[FileUploadCompleteResponse], await resp.json()
+        )
+        assert file_upload_complete_response.data  # nosec
+    state_url = file_upload_complete_response.data.links.state
+    log.info(
+        "completed upload of %s",
+        f"{upload_links=} with {parts=}, received {file_upload_complete_response=}",
+    )
 
-    data_provider = _file_sender(file_path, file_size)
-    headers = {"Content-Length": f"{file_size}"}
-    try:
-        async with session.put(url, data=data_provider, headers=headers) as resp:
-            if resp.status > 299:
-                response_text = await resp.text()
-                raise exceptions.S3TransferError(
-                    f"Could not upload file {file_path}:{response_text}"
+    log.debug("waiting for upload completion...")
+    async for attempt in AsyncRetrying(
+        reraise=True,
+        wait=wait_fixed(1),
+        stop=stop_after_delay(
+            NodePortsSettings.create_from_envs().NODE_PORTS_MULTIPART_UPLOAD_COMPLETION_TIMEOUT_S

```

Comment:
well for now this is a test value. AWS say that completion might take up to "several minutes". How do you interpret this?
I will set it to 15 minutes for now. and then we will test.

Reply:
and I do not understand your first question? can you elaborate?

Reply:
Since the NodePortSettings is used in different parts of the nodeports maybe we'd like to have it as a configuration parameter. Only instantiate it once at the initialisation of the library, like it was done for the RCloneSettings.